File: keyczar/src/com/google/keyczar/KeyczarTool.java
/*
 * Copyright 2008 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.keyczar;

import com.google.keyczar.enums.KeyPurpose;
import com.google.keyczar.enums.KeyStatus;
import com.google.keyczar.enums.KeyType;
import com.google.keyczar.exceptions.KeyczarException;

import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;


/**
 * Command line tool for generating Keyczar key files.
 * 
 * @author steveweis@gmail.com (Steve Weis)
 * @author arkajit.dey@gmail.com (Arkajit Dey)
 * 
 */
public class KeyczarTool {
  static String asymmetricFlag;
  static String destinationFlag;
  static String locationFlag;
  static String nameFlag;
  static KeyPurpose purposeFlag;
  static KeyStatus statusFlag = KeyStatus.ACTIVE;

  public static void main(String[] args) throws Exception {
    if (args.length == 0) {
      printUsage();
    } else {
      setFlags(args);
      if (args[0].equals("create")) {
        create();
      } else if (args[0].equals("addkey")) {
        addKey();
      } else if (args[0].equals("pubkey")) {
        publicKeys();
      } else if (args[0].equals("setstatus")) {
        setStatus();
      } else { // unsupported command
        printUsage();
      }
    }
  }

  private static void addKey() throws KeyczarException {
    // Read existing metadata
    if (locationFlag == null) {
      throw new KeyczarException("Must define a key set location with the "
          + "--location flag");
    }
    GenericKeyczar genericKeyczar = new GenericKeyczar(locationFlag);
    genericKeyczar.addVersion(statusFlag);
    genericKeyczar.write(locationFlag);
  }
  
  /**
   * Creates a new KeyMetadata object, deciding its name, purpose and type
   * based on command line flags. Outputs its JSON representation in a file 
   * named meta in the directory given by the --location flag.
   * 
   * @throws KeyczarException
   */
  private static void create() throws KeyczarException {
    KeyMetadata kmd = null;
    if (purposeFlag == null) {
      throw new KeyczarException("Must define a key set purpose with the "
          + "--purpose flag. Valid purposes are sign, crypt, and test.");
    }
    if (locationFlag == null) {
      throw new KeyczarException("Must define a key set location with the "
          + "--location flag.");
    }
    switch (purposeFlag) {
      case TEST:
        kmd = new KeyMetadata(nameFlag, KeyPurpose.TEST, KeyType.TEST);
        break;
      case SIGN_AND_VERIFY:
        if (asymmetricFlag != null) {
          if (asymmetricFlag.equalsIgnoreCase("rsa")) { // RSA
            kmd = new KeyMetadata(nameFlag, KeyPurpose.SIGN_AND_VERIFY,
                KeyType.RSA_PRIV);
          } else { // Default to DSA
            kmd = new KeyMetadata(nameFlag, KeyPurpose.SIGN_AND_VERIFY,
                KeyType.DSA_PRIV);
          }
        } else { // HMAC-SHA1
          kmd = new KeyMetadata(nameFlag, KeyPurpose.SIGN_AND_VERIFY,
              KeyType.HMAC_SHA1);
        }
        break;
      case DECRYPT_AND_ENCRYPT:
        if (asymmetricFlag != null) { // Default to RSA
          kmd = new KeyMetadata(nameFlag, KeyPurpose.DECRYPT_AND_ENCRYPT,
              KeyType.RSA_PRIV);
        } else { // AES
          kmd = new KeyMetadata(nameFlag, KeyPurpose.DECRYPT_AND_ENCRYPT,
              KeyType.AES);
        }
        break;
    }
    if (kmd == null) {
      throw new KeyczarException("Unsupported purpose: " + purposeFlag);
    }
    File file = new File(locationFlag + KeyczarFileReader.META_FILE);
    if (file.exists()) {
      throw new KeyczarException("File already exists: " + file);
    }

    try {
      FileOutputStream metaOutput = new FileOutputStream(file);
      metaOutput.write(kmd.toString().getBytes());
      metaOutput.close();
    } catch (IOException e) {
      throw new KeyczarException("Unable to write to : " + file.toString(), e);
    }
  }

  private static void printUsage() {
    // TODO: document --asymmetric flag, setstatus command
    String msg = "Usage:\t\"KeyczarTool command flags\"\n" +
                 "Commands:\n" + 
                 "create --name=name --location=location --purpose=purpose\n" +
                 "\tCreates a new key store\n" + 
                 "addkey --location=location --status=status\n" + 
                 "pubkey --location=location --destination=destination\n" +
                 "\tExport a key set at the given location as a set of " + 
                   "public keys at a given destination.\n" + 
                 "Flags:\n" + 
                 "\t--name : Define the name of a keystore. Optional.\n" + 
                 "\t--location : Define the file location of a keystore\n" + 
                 "\t--destination : Define the destination location of " + 
                   "a keystore.\n" + 
                 "\t--purpose : Define the purpose of a keystore." + 
                   " Must be sign, crypt, or test.\n" + 
                 "\t--status : Define the status of a new key. Must be " +
                   "primary, active, or scheduled_for_revocation. Optional.";
    System.out.println(msg);
  }

  private static void publicKeys() throws KeyczarException {
    if (locationFlag == null) {
      throw new KeyczarException("Must define a key set location with the "
          + "--location flag");
    }
    if (destinationFlag == null) {
      throw new KeyczarException("Must define a public key set location with"
          + " the --destination flag");
    }
    GenericKeyczar genericKeyczar = new GenericKeyczar(locationFlag);
    genericKeyczar.publicKeyExport(destinationFlag);
  }

  private static void setFlags(String[] args) {
    HashMap<String, String> params = new HashMap<String, String>();
    for (String arg : args) {
      if (arg.startsWith("--")) {
        arg = arg.substring(2); // Trim off the leading dashes
        String[] nameValuePair = arg.split("=");
        if (nameValuePair.length == 2) {
          params.put(nameValuePair[0], nameValuePair[1]);
        } else if (nameValuePair.length == 1) {
          params.put(nameValuePair[0], "true");
        }
      }
    }
    nameFlag = params.get("name");
    locationFlag = params.get("location");
    if (locationFlag != null && !locationFlag.endsWith(File.separator)) {
      locationFlag += File.separator;
    }

    destinationFlag = params.get("destination");
    if (destinationFlag != null && !destinationFlag.endsWith(File.separator)) {
      destinationFlag += File.separator;
    }

    purposeFlag = KeyPurpose.getPurpose(params.get("purpose"));
    //TODO: What about other purposes? ENCRYPT, VERIFY not included originally
    // also invalid purpose makes flag null, OK?
    statusFlag = KeyStatus.getStatus(params.get("status")); // default ACTIVE
    asymmetricFlag = params.get("asymmetric");
  }
  
  private static void setStatus() throws KeyczarException {
    if (locationFlag == null) {
      throw new KeyczarException("Must define a key set location with the "
          + "--location flag");
    }
    
    //TODO(arkajit.dey): finish implementing setStatus() method
  }

  private static class GenericKeyczar extends Keyczar {
    GenericKeyczar(String location) throws KeyczarException {
      super(location);
    }

    @Override
    boolean isAcceptablePurpose(KeyPurpose purpose) {
      return true;
    }

    void publicKeyExport(String destination) throws KeyczarException {
      KeyMetadata kmd = getMetadata();
      // Can only export if type is DSA_PRIV and purpose is SIGN_AND_VERIFY
      KeyMetadata publicKmd = null;
      if (kmd.getType() == KeyType.DSA_PRIV
          && kmd.getPurpose() == KeyPurpose.SIGN_AND_VERIFY) {
        publicKmd = new KeyMetadata(kmd.getName(), KeyPurpose.VERIFY,
            KeyType.DSA_PUB);
      } else if (kmd.getType() == KeyType.RSA_PRIV) {
        if (kmd.getPurpose() == KeyPurpose.DECRYPT_AND_ENCRYPT) {
          publicKmd = new KeyMetadata(kmd.getName(), KeyPurpose.ENCRYPT,
              KeyType.RSA_PUB);
        } else if (kmd.getPurpose() == KeyPurpose.SIGN_AND_VERIFY) {
          publicKmd = new KeyMetadata(kmd.getName(), KeyPurpose.VERIFY,
              KeyType.RSA_PUB);
        }
      }
      if (publicKmd == null) {
        throw new KeyczarException("Cannot export public keys for key type: "
            + kmd.getType() + " and purpose " + kmd.getPurpose());
      }
      Iterator<KeyVersion> versions = getVersions();
      while (versions.hasNext()) {
        KeyVersion version = versions.next();
        KeyczarKey publicKey = ((KeyczarPrivateKey) getKey(version))
            .getPublic();
        writeFile(publicKey.toString(), destination
            + version.getVersionNumber());
        publicKmd.addVersion(version);
      }
      writeFile(publicKmd.toString(), destination + KeyczarFileReader.META_FILE);
    }

    void write(String location) throws KeyczarException {
      writeFile(getMetadata().toString(), location
          + KeyczarFileReader.META_FILE);
      Iterator<KeyVersion> versions = getVersions();
      while (versions.hasNext()) {
        KeyVersion version = versions.next();
        writeFile(getKey(version).toString(), location
            + version.getVersionNumber());
      }
    }

    private void writeFile(String data, String location)
        throws KeyczarException {
      File outputFile = new File(location);
      try {
        FileWriter writer = new FileWriter(outputFile);
        writer.write(data);
        writer.close();
      } catch (IOException e) {
        throw new KeyczarException("Unable to write to : "
            + outputFile.toString(), e);
      }
    }
  }
}

File: cpp/src/keyczar/keyczar_tool.cc
// Copyright 2009 Sebastien Martini (seb@dbzteam.org)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
#include "keyczar/keyczar_tool.h"

#include <stdio.h>

#include "base/logging.h"
#include "base/string_util.h"
#include "base/sys_string_conversions.h"
#include "base/values.h"

#include "keyczar/key_type.h"
#include "keyczar/keyczar.h"
#include "keyczar/keyset.h"
#include "keyczar/keyset_encrypted_file_reader.h"
#include "keyczar/keyset_encrypted_file_writer.h"

namespace {

static const char kUsageMessage[] =
    "Usage: keyczart command flags\n"
    "Commands: create addkey pubkey promote demote revoke\n"
    "Flags: location name size status purpose destination version asymmetric\n"
    "       crypter\n\n"
    "Command usage:\n\n"
    "create --location=/path/to/keys --purpose=(crypt|sign) --name=\"A name\""
    " \\\n"
    "       --asymmetric=(dsa|rsa|ecdsa)\n"
    "   Creates a new, empty key set in the given location. This key set must\n"
    "   have a purpose of either \"crypt\" or \"sign\" and may optionally be\n"
    "   given a name. The optional asymmetric flag will generate a public \n"
    "   key set of the given algorithm. The \"dsa\" and \"ecdsa\" asymmetric\n"
    "   values are valid only for sets with \"sign\" purpose.\n\n"
    "addkey --location=/path/to/keys --status=(active|primary) --size=size \\\n"
    "       --crypter=crypterLocation\n"
    "   Adds a new key to an existing key set. Optionally specify a purpose,\n"
    "   which is \"active\" by default. Optionally specify a key size in\n"
    "   bits. Also optionally specify the location of a set of crypting keys,\n"
    "   which will be used to encrypt this key set.\n\n"
    "importkey --location=/path/to/keys --status=(active|primary) \\\n"
    "          --key=keyFileLocation --passphrase=\"passphrase\" \\\n"
    "          --crypter=crypterLocation\n"
    "   Imports a private key (RSA, DSA or ECDSA) to an existing key set.\n"
    "   keyFileLocation is the key file to import, it must have a PEM\n"
    "   representation. Optionally provide its passphrase. If a mandatory\n"
    "   passphrase is not specified as argument it will be prompted\n"
    "   interactively from the shell when this command is executed.\n"
    "   Optionally specify a purpose, which is \"active\" by default. Also\n"
    "   optionally the location of a set of crypting keys, which will be used\n"
    "   to encrypt this key set.\n\n"
    "pubkey --location=/path/to/keys --destination=/path/to/destination \\\n"
    "       --crypter=crypterLocation\n"
    "   Extracts public keys from a given key set and writes them to the\n"
    "   destination. The \"pubkey\" command only works for key sets that were\n"
    "   created with the \"--asymmetric\" flag.\n\n"
    "promote --location=/path/to/keys --version=versionNumber\n"
    "   Promotes the status of the given key version in the given location.\n"
    "   Active keys are promoted to primary (which demotes any existing\n"
    "   primary key to active). Keys scheduled for revocation are promoted to\n"
    "   be active.\n\n"
    "demote --location=/path/to/keys --version=versionNumber\n"
    "   Demotes the status of the given key version in the given location.\n"
    "   Primary keys are demoted to active. Active keys are scheduled for\n"
    "   revocation.\n\n"
    "revoke --location=/path/to/keys --version=versionNumber\n"
    "   Revokes the key of the given version number. This key must have been\n"
    "   scheduled for revocation by the promote command. WARNING: The key\n"
    "   will be destroyed.\n\n";

static void PrintUsage() {
  printf("%s", kUsageMessage);
}

static void MissingArgument(const std::wstring& argument) {
  LOG(ERROR) << "Missing argument '" << argument << "'.";
}

static bool GetSwitchValue(const CommandLine& command_line,
                           const std::wstring& switch_string,
                           std::string* value, bool optionnal) {
  if (value == NULL)
    return false;

  if (!command_line.HasSwitch(switch_string)) {
    if (!optionnal)
      MissingArgument(switch_string);
    return false;
  }

  value->assign(base::SysWideToUTF8(
                    command_line.GetSwitchValue(switch_string)));
  return true;
}

}  // namespace

namespace keyczar {

namespace keyczar_tool {

bool KeyczarTool::Init(int argc, char** argv) {
  // Initializes command line singleton
  CommandLine::Init(argc, argv);
  command_line_ = const_cast<CommandLine*>(CommandLine::ForCurrentProcess());
  if (command_line_ == NULL)
    return false;
  return true;
}

bool KeyczarTool::ProcessCommandLine() {
  if (command_line_->HasSwitch(L"help")) {
    PrintUsage();
    return true;
  }

  std::vector<std::wstring> loose_values = command_line_->GetLooseValues();
  if (loose_values.empty()) {
    LOG(ERROR) << "A command must be provided.";
    return false;
  }
  if (loose_values.size() > 1) {
    LOG(ERROR) << "A single command must be provided.";
    return false;
  }

  // A location is expected in every case.
  std::string location;
  if (!GetSwitchValue(*command_line_, L"location", &location, false))
    return false;

  // Command create
  if (loose_values[0] == L"create") {
    // purpose
    std::string purpose_string;
    if (!GetSwitchValue(*command_line_, L"purpose", &purpose_string, false))
      return false;
    scoped_ptr<KeyPurpose> purpose;
    if (purpose_string == "sign") {
      purpose.reset(KeyPurpose::Create("SIGN_AND_VERIFY"));
    } else {
      if (purpose_string == "crypt") {
        purpose.reset(KeyPurpose::Create("DECRYPT_AND_ENCRYPT"));
      } else {
        LOG(ERROR) << "Invalid purpose '" << purpose_string << "'.";
        return false;
      }
    }

    // name
    std::string name_string("Test");
    GetSwitchValue(*command_line_, L"name", &name_string, true);

    // asymmetric
    std::string asymmetric_string;
    GetSwitchValue(*command_line_, L"asymmetric", &asymmetric_string, true);

    return CmdCreate(location, *purpose, name_string, asymmetric_string);
  }

  // Command addkey
  if (loose_values[0] == L"addkey") {
    // status
    std::string status_string("ACTIVE");
    GetSwitchValue(*command_line_, L"status", &status_string, true);
    scoped_ptr<KeyStatus> status(KeyStatus::Create(
                                     StringToUpperASCII(status_string)));
    if (status.get() == NULL) {
      LOG(ERROR) << "Invalid status '" << status_string << "'.";
      return false;
    }

    // size
    std::string size_string;
    GetSwitchValue(*command_line_, L"size", &size_string, true);
    int size = 0;
    if (!size_string.empty() && !StringToInt(size_string, &size)) {
      LOG(ERROR) << "Invalid size '" << size_string << "'.";
      return false;
    }

    // crypter
    std::string crypter;
    GetSwitchValue(*command_line_, L"crypter", &crypter, true);

    return CmdAddKey(location, *status, size, crypter);
  }

  // Command importkey
  if (loose_values[0] == L"importkey") {
    // status
    std::string status_string("ACTIVE");
    GetSwitchValue(*command_line_, L"status", &status_string, true);
    scoped_ptr<KeyStatus> status(KeyStatus::Create(
                                     StringToUpperASCII(status_string)));
    if (status.get() == NULL) {
      LOG(ERROR) << "Invalid status '" << status_string << "'.";
      return false;
    }

    // filename
    std::string key_filename;
    if (!GetSwitchValue(*command_line_, L"key", &key_filename, false))
      return false;

    // passphrase
    std::string passphrase_string;
    std::string* passphrase = NULL;
    if (GetSwitchValue(*command_line_, L"passphrase", &passphrase_string, true))
      passphrase = &passphrase_string;

    // crypter
    std::string crypter;
    GetSwitchValue(*command_line_, L"crypter", &crypter, true);

    return CmdImportKey(location, *status, key_filename, passphrase, crypter);
  }

  // Command pubkey
  if (loose_values[0] == L"pubkey") {
    // destination
    std::string destination;
    if (!GetSwitchValue(*command_line_, L"destination", &destination, false))
      return false;

    // crypter
    std::string crypter;
    GetSwitchValue(*command_line_, L"crypter", &crypter, true);

    return CmdPubKey(location, destination, crypter);
  }

  // A version number is expected by all the remaining commands.
  std::string version_string;
  if (!GetSwitchValue(*command_line_, L"version", &version_string, false))
    return false;
  int version = 0;
  if (!StringToInt(version_string, &version)) {
    LOG(ERROR) << "Invalid version '" << version_string << "'.";
    return false;
  }

  // Command promote
  if (loose_values[0] == L"promote")
    return CmdPromote(location, version);
  // Command demote
  if (loose_values[0] == L"demote")
    return CmdDemote(location, version);
  // Command revoke
  if (loose_values[0] == L"revoke")
    return CmdRevoke(location, version);

  LOG(ERROR) << "Unknown command '" << loose_values[0] << "'.";
  PrintUsage();
  return false;
}

bool KeyczarTool::CmdCreate(const std::string& location,
                            const KeyPurpose& key_purpose,
                            const std::string& name,
                            const std::string& asymmetric) const {
  scoped_ptr<KeysetReader> reader(GetReader(location, ""));
  if (reader.get() == NULL)
    return false;

  scoped_ptr<Value> metadata_value(reader->ReadMetadata());
  if (metadata_value.get() != NULL) {
    LOG(ERROR) << "Metadata already exists, cannot overwrite.";
    return false;
  }
  metadata_value.reset();

  Keyset keyset;

  scoped_ptr<KeysetWriter> writer(GetWriter(location, ""));
  if (writer.get() == NULL)
    return false;

  keyset.AddObserver(writer.get());

  scoped_ptr<KeyType> key_type;
  if (key_purpose.type() == KeyPurpose::SIGN_AND_VERIFY) {
    if (asymmetric.empty()) {
#ifdef COMPAT_KEYCZAR_06B
      key_type.reset(KeyType::Create("HMAC_SHA1"));
#else
      key_type.reset(KeyType::Create("HMAC"));
#endif
    } else {
      if (asymmetric == "rsa") {
        key_type.reset(KeyType::Create("RSA_PRIV"));
      } else {
        if (asymmetric == "dsa") {
          key_type.reset(KeyType::Create("DSA_PRIV"));
        } else {
          if (asymmetric == "ecdsa") {
            key_type.reset(KeyType::Create("ECDSA_PRIV"));
          } else {
            LOG(ERROR) << "Invalid asymmetric argument '" << asymmetric << "'.";
            return false;
          }
        }
      }
    }
  } else {
    if (key_purpose.type() == KeyPurpose::DECRYPT_AND_ENCRYPT) {
      if (asymmetric.empty()) {
        key_type.reset(KeyType::Create("AES"));
      } else {
        if (asymmetric == "rsa") {
          key_type.reset(KeyType::Create("RSA_PRIV"));
        } else {
          LOG(ERROR) << "Invalid asymmetric argument '" << asymmetric << "'.";
          return false;
        }
      }
    } else {
      std::string purpose_string;
      key_purpose.GetName(&purpose_string);
      LOG(ERROR) << "Invalid key purpose '" << purpose_string << "'.";
      return false;
    }
  }

  KeysetMetadata* metadata = NULL;
  metadata = new KeysetMetadata(name, key_type.release(),
                                new KeyPurpose(key_purpose.type()),
                                false, 1);
  if (metadata == NULL)
    return false;

  keyset.set_metadata(metadata);
  return true;
}

bool KeyczarTool::CmdAddKey(const std::string& location,
                            const KeyStatus& key_status,
                            int size,
                            const std::string& crypter_location) const {
  scoped_ptr<KeysetReader> reader(GetReader(location, crypter_location));
  if (reader.get() == NULL)
    return false;

  scoped_ptr<Keyset> keyset(Keyset::Read(*reader, true));
  if (keyset.get() == NULL)
    return false;

  if (!crypter_location.empty() && !keyset->Empty() && keyset->metadata() &&
      !keyset->metadata()->encrypted()) {
    LOG(ERROR) << "This is a non encrypted key set which already contains "
               << "non encrypted keys.";
    return false;
  }

  scoped_ptr<KeysetWriter> writer(GetWriter(location, crypter_location));
  if (writer.get() == NULL)
    return false;

  keyset->AddObserver(writer.get());

  if (size == 0) {
    if (keyset->GenerateDefaultKeySize(key_status.type()) == 0)
      return false;
  }  else {
    if (keyset->GenerateKey(key_status.type(), size) == 0)
      return false;
  }

  if (!crypter_location.empty())
    keyset->set_encrypted(true);
  return true;
}

bool KeyczarTool::CmdImportKey(const std::string& location,
                               const KeyStatus& key_status,
                               const std::string& filename,
                               const std::string* passphrase,
                               const std::string& crypter_location) const {
  scoped_ptr<KeysetReader> reader(GetReader(location, crypter_location));
  if (reader.get() == NULL)
    return false;

  scoped_ptr<Keyset> keyset(Keyset::Read(*reader, true));
  if (keyset.get() == NULL)
    return false;

  if (!crypter_location.empty() && !keyset->Empty() && keyset->metadata() &&
      !keyset->metadata()->encrypted()) {
    LOG(ERROR) << "This is a non encrypted key set which already contains "
               << "non encrypted keys.";
    return false;
  }

  scoped_ptr<KeysetWriter> writer(GetWriter(location, crypter_location));
  if (writer.get() == NULL)
    return false;

  keyset->AddObserver(writer.get());

  if (keyset->ImportPEMKey(key_status.type(), filename, passphrase) == 0)
    return false;

  if (!crypter_location.empty())
    keyset->set_encrypted(true);
  return true;
}

bool KeyczarTool::CmdPubKey(const std::string& location,
                            const std::string& destination,
                            const std::string& crypter_location) const {
  scoped_ptr<KeysetReader> reader(GetReader(location, crypter_location));
  if (reader.get() == NULL)
    return false;

  scoped_ptr<Keyset> keyset(Keyset::Read(*reader, true));
  if (keyset.get() == NULL)
    return false;

  scoped_ptr<KeysetWriter> writer(GetWriter(destination, ""));
  if (writer.get() == NULL)
    return false;

  return keyset->PublicKeyExport(*writer);
}

bool KeyczarTool::CmdPromote(const std::string& location, int version) const {
  scoped_ptr<KeysetReader> reader(GetReader(location, ""));
  if (reader.get() == NULL)
    return false;

  scoped_ptr<Keyset> keyset(Keyset::ReadMetadataOnly(*reader));
  if (keyset.get() == NULL)
    return false;

  scoped_ptr<KeysetWriter> writer(GetWriter(location, ""));
  if (writer.get() == NULL)
    return false;

  keyset->AddObserver(writer.get());

  return keyset->PromoteKey(version);
}

bool KeyczarTool::CmdDemote(const std::string& location, int version) const {
  scoped_ptr<KeysetReader> reader(GetReader(location, ""));
  if (reader.get() == NULL)
    return false;

  scoped_ptr<Keyset> keyset(Keyset::ReadMetadataOnly(*reader));
  if (keyset.get() == NULL)
    return false;

  scoped_ptr<KeysetWriter> writer(GetWriter(location, ""));
  if (writer.get() == NULL)
    return false;

  keyset->AddObserver(writer.get());

  return keyset->DemoteKey(version);
}

bool KeyczarTool::CmdRevoke(const std::string& location, int version) const {
  scoped_ptr<KeysetReader> reader(GetReader(location, ""));
  if (reader.get() == NULL)
    return false;

  scoped_ptr<Keyset> keyset(Keyset::ReadMetadataOnly(*reader));
  if (keyset.get() == NULL)
    return false;

  scoped_ptr<KeysetWriter> writer(GetWriter(location, ""));
  if (writer.get() == NULL)
    return false;

  keyset->AddObserver(writer.get());

  // TODO(seb): currently the key version is not erased from disk.
  return keyset->RevokeKey(version);
}

KeysetReader* KeyczarTool::GetReader(
    const std::string& location,
    const std::string& crypter_location) const {
  switch (location_type_) {
    case DISK:
      return GetFileReader(location, crypter_location);
    default:
      NOTREACHED();
  }

  return NULL;
}

KeysetWriter* KeyczarTool::GetWriter(
    const std::string& location,
    const std::string& crypter_location) const {
  switch (location_type_) {
    case DISK:
      return GetFileWriter(location, crypter_location);
    default:
      NOTREACHED();
  }

  return NULL;
}

KeysetFileReader* KeyczarTool::GetFileReader(
    const std::string& location, const std::string& crypter_location) const {
  scoped_ptr<KeysetFileReader> reader;
  scoped_ptr<Crypter> crypter;

  if (crypter_location.empty()) {
    reader.reset(new KeysetFileReader(location));
  } else {
    crypter.reset(Crypter::Read(crypter_location));
    if (crypter.get() == NULL)
      return NULL;
    reader.reset(new KeysetEncryptedFileReader(location, crypter.release()));
  }
  return reader.release();
}

KeysetFileWriter* KeyczarTool::GetFileWriter(
    const std::string& location, const std::string& crypter_location) const {
  scoped_ptr<KeysetFileWriter> writer;
  scoped_ptr<Encrypter> encrypter;

  if (crypter_location.empty()) {
    writer.reset(new KeysetFileWriter(location));
  } else {
    encrypter.reset(Encrypter::Read(crypter_location));
    if (encrypter.get() == NULL)
      return false;

    writer.reset(new KeysetEncryptedFileWriter(location, encrypter.release()));
  }
  return writer.release();
}

}  // namespace keyczar_tool

}  // namespace keyczar
